torch/csrc/distributed/c10d/c10d_frontend.h

```diff
@@ -39,180 +39,184 @@ class DistributedC10d {
       at::Tensor tensor,
       int64_t src,
-      std::shared_ptr<ProcessGroup> group,
-      bool async_op);
+      const std::shared_ptr<ProcessGroup>& group,
+      bool async_op = false);
 
-  c10::optional<ProcessGroup::Work> allReduceMultiGPU(
+  std::shared_ptr<ProcessGroup::Work> allReduceMultiGPU(
       std::vector<at::Tensor>& tensor_list,
-      ReduceOp op,
-      std::shared_ptr<ProcessGroup> group,
-      bool async_op);
+      const std::shared_ptr<ProcessGroup>& group,
+      ReduceOp op = ReduceOp::SUM,
+      bool async_op = false);
 
-  c10::optional<ProcessGroup::Work> allReduce(
+  std::shared_ptr<ProcessGroup::Work> allReduce(
       at::Tensor tensor,
-      ReduceOp op,
-      std::shred_ptr<ProcessGroup> group,
-      bool async_op);
+      const std::shared_ptr<ProcessGroup>& group,
+      ReduceOp op = ReduceOp::SUM,
+      bool async_op = false);
 
-  c10::optional<ProcessGroup::Work> allReduceCoalesced(
-      at::Tensor tensor,
-      ReduceOp op,
-      std::shred_ptr<ProcessGroup> group,
-      bool async_op);
+  std::shared_ptr<ProcessGroup::Work> allReduceCoalesced(
+      std::vector<at::Tensor>& tensors,
+      const std::shared_ptr<ProcessGroup>& group,
+      ReduceOp op = ReduceOp::SUM,
+      bool async_op = false);
 
-  c10::optional<ProcessGroup::Work> reduceMultiGPU(
+  std::shared_ptr<ProcessGroup::Work> reduceMultiGPU(
       std::vector<at::Tensor>& tensor_list,
       int64_t dst,
-      ReduceOp op,
-      std::shared_ptr<ProcessGroup> group,
-      bool async_op,
-      int64_t dst_tensor);
+      const std::shared_ptr<ProcessGroup>& group,
+      ReduceOp op = ReduceOp::SUM,
+      bool async_op = false,
+      int64_t dst_tensor = 0);
 
-  c10::optional<ProcessGroup::Work> reduce(
+  std::shared_ptr<ProcessGroup::Work> reduce(
       at::Tensor tensor,
       int64_t dst,
-      ReduceOp op,
-      std::shared_ptr<ProcessGroup>& group,
-      bool async_op);
+      const std::shared_ptr<ProcessGroup>& group,
+      ReduceOp op = ReduceOp::SUM,
+      bool async_op = false);
 
-  c10::optional<ProcessGroup::Work> allGatherMultiGPU(
+  std::shared_ptr<ProcessGroup::Work> allGatherMultiGPU(
       std::vector<std::vector<at::Tensor>>& output_tensor_lists,
-      const std::vector<at::Tensor>& input_tensor_list,
-      std::shared_ptr<ProcessGroup> group,
-      bool async_op);
+      std::vector<at::Tensor>& input_tensor_list,
+      const std::shared_ptr<ProcessGroup>& group,
+      bool async_op = false);
 
-    // TODO TODO following APIs take python objects and unpickle them, how do we support these?
-    // ProcessGroup::Work allGatherObject()
-    // ProcessGroup::Work gatherObject()
-    // ProcessGroup::Work broadcastObjectList()
+  // TODO TODO following APIs take python objects and unpickle them, how do we
+  // support these? ProcessGroup::Work allGatherObject() ProcessGroup::Work
+  // gatherObject() ProcessGroup::Work broadcastObjectList()
```

Comment:
maybe we can remove those TODOs now as we discussed we would like to keep them in python?

Reply:
Sure. done.